File: JavaScript/src/Internals/Functions/MathSum2/Function_TDIST.js
import { Function_3 } from '../Function_3.js';
import { Operand } from '../../../Operand.js';
import { ExcelFunctions } from '../../../MathNet/ExcelFunctions.js';

class Function_TDIST extends Function_3 {
    get Name() {
        return "TDist";
    }

    constructor(z) {
        super(z);
    }

    evaluate(engine, tempParameter) {
        let args1 = this.getNumber_1(engine, tempParameter);
        if (args1.IsError) return args1;

        let args2 = this.getNumber_2(engine, tempParameter);
        if (args2.IsError) return args2;

        let args3 = this.getNumber_3(engine, tempParameter);
        if (args3.IsError) return args3;
        let x = args1.DoubleValue;
        let degreesFreedom = args2.IntValue;
        let tails = args3.IntValue;
        if (degreesFreedom <= 0.0 || tails < 1 || tails > 2) {
            return this.functionError();
        }
        return Operand.Create(ExcelFunctions.TDist(x, degreesFreedom, tails));
    }
}

export { Function_TDIST };

File: java/toolgood.algorithm/src/main/java/toolgood/algorithm/internals/functions/mathSum2/Function_TDIST.java
package toolgood.algorithm.internals.functions.mathSum2;

import java.math.BigDecimal;
import java.util.List;
import java.util.function.BiFunction;
import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.NoneEngine;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_3;
import toolgood.algorithm.mathNet.ExcelFunctions;

final class Function_TDIST extends Function_3 {
    public Function_TDIST(FunctionBase[] funcs) {
        super(funcs);
        if (funcs.length != 3) {
            throw new IllegalArgumentException("Function '" + Name() + "' requires exactly 3 parameters.");
        }
    }

    @Override
    public String Name() { return "TDist"; }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        Operand args1 = GetNumber_1(engine, tempParameter);
        if (args1.IsErrorOrNone()) return args1;

        Operand args2 = GetNumber_2(engine, tempParameter);
        if (args2.IsErrorOrNone()) return args2;

        Operand args3 = GetNumber_3(engine, tempParameter);
        if (args3.IsErrorOrNone()) return args3;

        int degreesFreedom = args2.IntValue();
        if (degreesFreedom <= 0) {
            return ParameterError(2);
        }
        int tails = args3.IntValue();
        if (tails < 1 || tails > 2) {
            return ParameterError(3);
        }
        try {
            return Operand.Create(BigDecimal.valueOf(ExcelFunctions.TDist(args1.NumberValue().doubleValue(), degreesFreedom, tails)));
        } catch (Exception e) {
            return FunctionError();
        }
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.NUMBER;
    }

    @Override
    void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, String op, String val) {
        func1.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
        func2.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
        func3.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
    }
}
